static/panes/compiler.ts

```diff
@@ -0,0 +1,3561 @@
+            if (this.compiler.notification) {
+                this.compiler.notification.forEach(notification =>
+                    this.alertSystem.notify(notification, {
+                        group: 'compilerwarning',
+                        alertClass: 'notification-info',
+                        dismissTime: 7000,
+                    })
+                );
+            }
+            this.prependOptions.data('content', this.compiler.options);
+        }
+    }
+
+    updateCompilerUI(): void {
+        this.initToolButtons();
+        this.updateButtons();
+        this.updateCompilerInfo();
+        // Resize in case the new compiler name is too big
+        this.resize();
+    }
+
+    onCompilerChange(value: string): void {
+        this.compiler = this.compilerService.findCompiler(this.currentLangId ?? '', value);
+
+        this.deferCompiles = true;
+        this.needsCompile = true;
+
+        this.updateLibraries();
+        this.saveState();
+        this.updateCompilerUI();
+
+        this.undefer();
+
+        this.sendCompiler();
+    }
+
+    sendCompiler(): void {
+        this.eventHub.emit(
+            'compiler',
+            this.id,
+            this.compiler,
+            this.options as any,
+            this.sourceEditorId ?? 0,
+            this.sourceTreeId ?? 0
+        );
+    }
+
+    onEditorClose(editor: number): void {
+        if (editor === this.sourceEditorId) {
+            // We can't immediately close as an outer loop somewhere in GoldenLayout is iterating over
+            // the hierarchy. We can't modify while it's being iterated over.
+            this.close();
+            _.defer(() => {
+                this.container.close();
+            });
+        }
+    }
+
+    onTreeClose(tree: any): void {
+        if (tree === this.sourceTreeId) {
+            this.close();
+            _.defer(() => {
+                this.container.close();
+            });
+        }
+    }
+
+    onFilterChange(): void {
+        const filters = this.getEffectiveFilters();
+        this.eventHub.emit('filtersChange', this.id, filters as any);
+        this.saveState();
+        this.compile();
+        this.updateButtons();
+    }
+
+    currentState(): CurrentStateType {
+        const state = {
+            id: this.id,
+            compiler: this.compiler ? this.compiler.id : '',
+            source: this.sourceEditorId,
+            tree: this.sourceTreeId,
+            options: this.options,
+            // NB must *not* be effective filters
+            filters: this.filters.get(),
+            wantOptInfo: this.wantOptInfo,
+            libs: this.libsWidget?.get(),
+            lang: this.currentLangId,
+            selection: this.selection,
+            flagsViewOpen: this.flagsViewOpen,
+        };
+        this.paneRenaming.addState(state);
+        this.fontScale.addState(state);
+        return state;
+    }
+
+    saveState(): void {
```

Comment:
`saveState` isn't needed, it's the same as the Pane's `updateState` (except `updateState` calls `getCurrentState`)